dask/distributed/scheduler.py

```diff
@@ -717,5 +717,33 @@ def _get_workers(self, header, payload):
     def _heartbeat(self, header, payload):
         with logerrors():
             log(self.address_to_clients, "Heartbeat", header)
+            payload = pickle.loads(payload)
             address = header['address']
+
+            if address not in self.workers:
+                self.available_workers.put(address)
+
+            self.workers[address] = payload
             self.workers[address]['last-seen'] = datetime.utcnow()
+
+    def prune_workers(self, timeout=20):
+        """
+        Remove workers from scheduler that have not sent a heartbeat in
+        `timeout` seconds.
+        """
+        now = datetime.utcnow()
+        remove = []
+        with self.lock:
```

Comment:
Yes, the lock is a normal threading lock.  In practice it has so far been used to isolate the zeromq sockets (although this should ideally be unnecessary if they only operate in the same thread).

Generally one associates a lock to a set of resources that can not be manipulated at the same time.  The zmq sockets are one such resource set.  I'm not sure if you intend to include `self.workers` in this same resource set or not.